build/utf8/valid_asm.go

```diff
@@ -0,0 +1,547 @@
+
+		// ____0100 ________
+		Carry | TooLarge,
+		// ____0101 ________
+		Carry | TooLarge | TooLarge1000,
+		// ____011_ ________
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+
+		// ____1___ ________
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+		// ____1101 ________
+		Carry | TooLarge | TooLarge1000 | Surrogate,
+		Carry | TooLarge | TooLarge1000,
+		Carry | TooLarge | TooLarge1000,
+	})
+
+	nib3 = fullMask([16]byte{
+		// ________ 0_______ <ASCII in byte 2>
+		TooShort, TooShort, TooShort, TooShort,
+		TooShort, TooShort, TooShort, TooShort,
+
+		// ________ 1000____
+		TooLong | Overlong2 | TwoConts | Overlong3 | TooLarge1000 | Overlong4,
+		// ________ 1001____
+		TooLong | Overlong2 | TwoConts | Overlong3 | TooLarge,
+		// ________ 101_____
+		TooLong | Overlong2 | TwoConts | Surrogate | TooLarge,
+		TooLong | Overlong2 | TwoConts | Surrogate | TooLarge,
+
+		// ________ 11______
+		TooShort, TooShort, TooShort, TooShort,
+	})
+
+	return
+}
+
+func main() {
+	TEXT("Valid", NOSPLIT, "func(p []byte) bool")
+	Doc("Valid reports whether p consists entirely of valid UTF-8-encoded runes.")
+
+	ret, _ := ReturnIndex(0).Resolve()
+
+	d := Load(Param("p").Base(), GP64())
+	n := Load(Param("p").Len(), GP64())
+
+	JumpUnlessFeature("stdlib", cpu.AVX2)
+
+	// 128 has been found empirically on my machine. After that size, the
+	// AVX2 implementation is faster than the stdlib one.
+	Comment("if input < 128 bytes")
+	CMPQ(n, U8(128))
+	JGE(LabelRef("init_avx"))
+
+	Label("stdlib")
+	stdlib(d, n, ret)
+
+	Label("init_avx")
+
+	scratch := AllocLocal(32)
+	scratchAddr := GP64()
+	LEAQ(scratch, scratchAddr)
+
+	Comment("Prepare the constant masks")
+
+	incompleteMask := ConstBytes("incomplete_mask", incompleteMaskData())
+	incompleteMaskY := YMM()
+	VMOVDQU(incompleteMask, incompleteMaskY)
+
+	continuation4Bytes := ConstBytes("cont4_vec", continuationMaskData(0b11110000))
+	continuation4BytesY := YMM()
+	VMOVDQU(continuation4Bytes, continuation4BytesY)
+
+	continuation3Bytes := ConstBytes("cont3_vec", continuationMaskData(0b11100000))
+	continuation3BytesY := YMM()
+	VMOVDQU(continuation3Bytes, continuation3BytesY)
+
+	nib1Data, nib2Data, nib3Data := nibbleMasksData()
+
+	Comment("High nibble of current byte")
+	nibble1Errors := ConstBytes("nibble1_errors", nib1Data)
+	nibble1Y := YMM()
+	VMOVDQU(nibble1Errors, nibble1Y)
+
+	Comment("Low nibble of current byte")
+	nibble2Errors := ConstBytes("nibble2_errors", nib2Data)
+	nibble2Y := YMM()
+	VMOVDQU(nibble2Errors, nibble2Y)
+
+	Comment("High nibble of the next byte")
+	nibble3Errors := ConstBytes("nibble3_errors", nib3Data)
+	nibble3Y := YMM()
+	VMOVDQU(nibble3Errors, nibble3Y)
+
+	Comment("Nibble mask")
+	nibbleMask := ConstArray64("nibble_mask",
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+		0x0F0F0F0F0F0F0F0F,
+	)
```

Comment:
Do you mind add a comment or link that documents how this mask was picked?

Reply:
I thought the name made it clear it was a mask to get each (lower) nibble. Will add a comment.

Reply:
Maybe rename to `lower_nibble_mask` ? :)